src/hotspot/share/c1/c1_GraphBuilder.cpp

```diff
@@ -2133,11 +2133,23 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
     if ((code == Bytecodes::_invokestatic && klass->is_initialized()) || // invokestatic involves an initialization barrier on declaring class
         code == Bytecodes::_invokespecial ||
         (code == Bytecodes::_invokevirtual && target->is_final_method()) ||
-        code == Bytecodes::_invokedynamic) {
+        code == Bytecodes::_invokedynamic ||
+        target->get_Method()->intrinsic_id() == vmIntrinsics::_clone) {
       // static binding => check if callee is ok
       ciMethod* inline_target = (cha_monomorphic_target != nullptr) ? cha_monomorphic_target : target;
       bool holder_known = (cha_monomorphic_target != nullptr) || (exact_target != nullptr);
-      bool success = try_inline(inline_target, holder_known, false /* ignore_return */, code, better_receiver);
+      bool success;
+      // Clone intrinsic and inlining can only kick when instance is a primitive array,
+      // or when the target of the clone is not a Phi node
+      ciType* receiver_type;
+      if (target->get_Method()->intrinsic_id() == vmIntrinsics::_clone &&
+          ((receiver_type = state()->stack_at(state()->stack_size() - inline_target->arg_size())->exact_type()) == nullptr || // clone target is phi
```

Comment:
I don't think target-specific logic belongs here.  And I don't understand the point about Phi nodes.  Isn't the holder_known flag enough?  For primitive arrays, isn't it true that inline_target->get_Method()->intrinsic_id() == vmIntrinsics::_clone?

Reply:
> I don't think target-specific logic belongs here. And I don't understand the point about Phi nodes. Isn't the holder_known flag enough?

In my testing `holder_known` was not enough to detect objects that are not Phi. For example:

```
    static int[] test(int[] ints)
    {
        return ints.clone();
    }
```

`holder_known` is false when it tries to C1 compile `ints.clone()`, am I missing something here?

> For primitive arrays, isn't it true that inline_target->get_Method()->intrinsic_id() == vmIntrinsics::_clone?

Possibly, but in this part of the logic I'm trying to find situations in which I don't want to apply the `clone` intrinsic. And those situations are non-array objects, and for arrays, those whose elements are not primitives. I don't see how I can craft such a condition with only `inline_target->get_Method()->intrinsic_id() == vmIntrinsics::_clone`? IOW, that condition might be true for primitive arrays, but is it false for non-array objects and non-primitive arrays?

Reply:
You're right about holder_known, but why do you need to check for _clone specifically at line 2137?  If there is logic missing that prevents an inlining attempt then I think it should be fixed first, rather than in a followup.

And I see that you need to do a receiver type check to allow only primitive arrays.  Can you do that in append_alloc_array_copy, and bailout if not successful?  The logic in build_graph_for_intrinsic would need to change slightly to support this.

Reply:
I was able to remove the clone-specific logic in invoke() in two parts:

1. fix the type_is_exact logic to allow array receiver
2. move primitive array receiver check into append_alloc_array_copy

Reply:
> You're right about holder_known, but why do you need to check for _clone specifically at line 2137? If there is logic missing that prevents an inlining attempt then I think it should be fixed first, rather than in a followup.

I added that check because none of the conditions in that `if` statement satisfied the situations in which `clone` calls are optimized. For the example I gave above, `code == Bytecodes::_invokevirtual` is true and `target->is_final_method()` is false. So that's why I added `clone` specifically.

> I was able to remove the clone-specific logic in invoke() in two parts:
> 
> 1. fix the type_is_exact logic to allow array receiver
> 2. move primitive array receiver check into append_alloc_array_copy

Great! I assume you also solved the clone check in line 2137?

How do we add your work on top of mine? Do I cherry pick the commit(s) from a branch of yours? Or some other way?